### ISSUE TITLE
Base tile layers fetched when invisible

### ISSUE THREAD

**Issue by garthk** [Leaflet/Leaflet#4925]
~~Is there a way to not fetch tiles for tile layers that aren't visible?~~

Desired behaviour:

* Documented means to have Leaflet download tiles for only the selected base layer according to `L.control.layers`

Current behaviour (0.7.7):

* Leaflet downloads tiles for all layers, whether visible according to `L.control.layers` or not
* Tiles for the "invisible" layer can appear before the visible tiles if they load first, irritating the users 
* The documentation does not mention any way to change this
* I can't find any undocumented means in the code, either

To reproduce:

* `L.tileLayer('https://api.tiles.mapbox.com/v4/{id}/{z}/{x}/{y}.png?access_token={accessToken}', options).addTo(map)` with your `accessToken` once each for `id: 'mapbox.satellite'` and `id: 'mapbox.outdoors'` to produce `satelliteTileLayer` and `outdoorsTileLayer`
* `L.control.layers({ sat: satelliteTileLayer, map: outdoorsTileLayer })`
* Scroll around at zoom 14
* Observe satellite tiles visibly loading before outdoors tiles load

**Comment by IvanSanchez** [comment]
The answer is probably [yes](http://www.strcat.de/questions-with-yes-or-no-answers.html).

However, this issue tracker is used for reporting bugs and discussing new features. For questions on using Leaflet, please use [gis.stackexchange.com](https://gis.stackexchange.com/questions/tagged/leaflet) or [stackoverflow](https://stackoverflow.com/questions/tagged/leaflet).

**Comment by garthk** [comment]
Rephrased; see above.

If the feature exists, it's not documented, so it's a documentation bug.

If the feature does not exist, it's a feature request.

Please re-open, per your advice that this issue tracker is used for reporting bugs and discussing new features.

**Comment by IvanSanchez** [comment]
> Leaflet downloads tiles for all layers, whether visible according to `L.control.layers` or not

No, it doesn't. See https://playground-leaflet.rhcloud.com/qog/edit?html,output

If you disagree, please publish a fiddle/codepen/[playground](https://playground-leaflet.rhcloud.com/) that displays this behaviour.

**Comment by garthk** [comment]
Thanks for the demo; I could diff it against my code to isolate the problem with my usage pattern. Amend your example as follows to see it:

```js
    // insert these:
    map.addLayer(positron);
    map.addLayer(darkmatter);
    // before these:
    L.control.layers({Positron: positron, 'Dark Matter': darkmatter}, {}, {collapsed: false}).addTo(map);
```

_Et voilà_, double loading as claimed.

The relevant chunk of documentation, under [Control Layers](http://leafletjs.com/reference.html#control-layers), emphasis mine:

> Creates an attribution control with the given layers. Base layers will be switched with radio buttons, while overlays will be switched with checkboxes. **Note that all base layers should be passed in the base layers object, but only one should be added to the map during map instantiation.**

The blunder is obvious in hindsight. My bad. And, we'll be friendlier if we guard rail against it. I'll dig for a way to identify the double-add and either a) warn against it, or b) quietly suppress it. If I can't find one, a stronger warning in the docs might suffice.

Is there a convention in the code base for issuing warnings? I don't see any use of `console`.

**Comment by garthk** [comment]
Is this within cooee of acceptable?

```js
	onAdd: function (map) {
		this._initLayout();
		this._update();

		map
		    .on('layeradd', this._onLayerChange, this)
		    .on('layerremove', this._onLayerChange, this);

		// Unload redundant base layers under our control
		// https://github.com/Leaflet/Leaflet/issues/4925
		var alreadyLoadedBaseLayerCount = 0;
		for (var i in this._layers) {
			var entry = this._layers[i];
			if (!entry.overlay && map.hasLayer(entry.layer)) {
				if (++alreadyLoadedBaseLayerCount > 1) {
					map.removeLayer(entry.layer);
				}
			}
		}

		return this._container;
	},
```

Roughly equivalent, but leaving the last-mentioned base layer selected rather than the first-mentioned:

```js
		// Unload redundant base layers under our control
		// https://github.com/Leaflet/Leaflet/issues/4925
		this._onInputClick();
``` 

`npm test` passes for both.

**Comment by IvanSanchez** [comment]
> Is there a convention in the code base for issuing warnings? I don't see any use of `console`.

We still haven't dropped support for IE8, so we don't assume `console`.

I kinda like calling `this._onInputClick();`, but I'm not 100% sure about it. It makes sense, and looks like it would guard against developer mishaps. Maybe make it optional (or even default) e.g. `L.control.layers(bases, overs, {sanitizeLayers: false})`. And if this is going to be a feature, there should be tests for it.

Such a change should also be noted in `docs/examples/layers-control.md`
